### PR TITLE
Documentation changes

### PR DESCRIPTION
**Pull request by vtantia**
Intermediate work while creating documentation

### CODE REVIEW

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 1]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+
+        slowmo_momentum (float): This specifies the value of slowmo momentum to be used (read
+                        https://arxiv.org/abs/1910.00643 for more details). This parameter
+                        might need to be tuned and the optimal value varies according to
+                        the use case and the number of nodes being run on. The optimal value
+                        typically increases with the number of nodes. On training
+                        transfomers on the WMT 16 En-De dataset, we have found the optimal
+                        values to be 0 for less than 4 nodes, 0.2 for 4 nodes, 0.5 for 8
+                        nodes and 0.6 for 16 nodes (default: 0.5)
+        slowmo_frequency (int): This specifies how often slow momentum is to be performed.
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:142]: I would remove the line saying "This does not need to be tuned generally" because I think it really does depend on the application. (For other applications, we've been using much smaller values, around 8-12.)

Also, after "how often" you could specify the units to make it more precise: "This specifies how often (iterations) slow momentum..."

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 2]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+                         This does not need to be tuned generally (default: 48)
+        slowmo_lr (float): This specifies the value of slowmo learning rate to be used (read
+                  https://arxiv.org/abs/1910.00643 for more details). This does not need to
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:145]: Rather than "This does not need to be tuned generally." I'd suggest changing to the stronger wording "We do not recommend changing this."

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 3]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+        and others whose format is ``torch.channels_last``.  However,
+        corresponding parameters in different processes must have the
+        same strides.
+    .. warning::
+        Forward and backward hooks defined on :attr:`module` and its submodules
+        won't be invoked anymore, unless the hooks are initialized in the
+        :meth:`forward` method.
+    Args:
+        module (Module): module to be parallelized
+        nprocs_per_node (int): Number of GPUs per node. This needs to be specified for
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:116]: In the code this has a default value, but it seems like it may be better to remove the default value and force the user to specify it. What do you think? (No default is mentioned in the documentation.)

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 4]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+                        typically increases with the number of nodes. On training
+                        transfomers on the WMT 16 En-De dataset, we have found the optimal
+                        values to be 0 for less than 4 nodes, 0.2 for 4 nodes, 0.5 for 8
+                        nodes and 0.6 for 16 nodes (default: 0.5)
+        slowmo_frequency (int): This specifies how often slow momentum is to be performed.
+                         This does not need to be tuned generally (default: 48)
+        slowmo_lr (float): This specifies the value of slowmo learning rate to be used (read
+                  https://arxiv.org/abs/1910.00643 for more details). This does not need to
+                  be tuned generally (default: 1.0)
+        slowmo_num_shards (int): The number of shards between which slow momentum parameters
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:147]: I worry that the description here may not be sufficient for a new user to understand this functionality.

Also, I looked through the code quickly, but I think I'm missing something. Is there a default setting where we don't do any sharding (so each process keeps a full copy of the slowmo state? I would expect that to be the default (how the algorithm was originally described in the paper), and we only shard if the user actually species to do so. (Using sharding reduces the memory footprint but increases communication overhead, right?)

Right now the default to use 32 shards seems pretty arbitrary.

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 5]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+                           (default: 3)
+
+        process_group: The process group to be used for distributed data
+                       all-reduction. If ``None``, the default process group, which
+                       is created by :func:`torch.distributed.init_process_group`,
+                       will be used. (default: ``None``)
+
+        SGP parameters
+
+        graph: Optional[GraphManager] = None,
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:163]: I guess the documentation for these parameters still needs to be added?

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 6]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+        use_streams: bool = True,
+        slowmo_sgp_average_params: bool = False,
+
+        Debugging
+
+        verbose (bool): prints various logs which are useful for debugging (default: False)
+        profile_mode (bool): prints the time taken by different parts of the code, which can help
+                     in finding bottlenecks (default: False)
+
+        Only for advanced users
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:177]: Could add a note here that if these are not specified, the default behavior is to get them from the torch.distributed module

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 7]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+                     by init_process_group (default: None)
+        master_group (Optional[torch.distributed.ProcessGroup]): Process group which only contains
+                     the master GPUs of each node (default: None)
+        local_node_group (Optional[torch.distributed.ProcessGroup]): Process group which only
+                         contains the GPUs local to the current node (default: None)
+        comm_device: (Optional[torch.device]): The torch.device on which torch tensors are to be
+                     placed before communication (default: None)
+    Example:
+        >>> torch.distributed.init_process_group(backend='nccl', world_size=4, init_method='...')
+        >>> net = fairscale.data_parallel.GossipDataParallel(model, pg, nprocs_per_node)
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:191]: What is `pg` in this example?

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 8]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+        >>> loss = criterion(net(inputs), targets)
+        >>> loss.backward()
+        >>> optimizer.step()
+        >>> net.perform_slowmo(optimizer)
+        If fp16 training is being done, the last line changes to
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/distributed.py:196]: Formatting looks weird, a line of text (comment?) in the middle of lines of code

File: fairscale/nn/data_parallel/gossip/utils/helpers.py [hunk 9]
```diff
@@ -99,11 +97,11 @@ def communicate(tensors: List[torch.Tensor], communication_op: Any, logger: logg
 
 HANDLER_AND_LEVEL_SET: Set[logging.Logger] = set()
 
-# TODO: need to deprecate this function
+# TODO: deprecate this function
```
Comment (mikerabbat) [inline fairscale/nn/data_parallel/gossip/utils/helpers.py:100]: Is this function still used in the current implementation? If not, may as well clean it up before sending the PR

**Review by mikerabbat — CHANGES_REQUESTED** [review]
Please see the inline comments for some questions.

Does the slowmo implementation allow for doing no sharding (i.e., keeping and updating a full copy of the slowmo state at every worker, as was done in the original SlowMo paper?

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 10]
```diff
@@ -53,22 +53,160 @@ class SlowmoBaseAlgorithm(Enum):
...
+    splitting the input across the specified devices by chunking in the batch
+    dimension. The module is replicated on each machine and each device, and
+    each such replica handles a portion of the input. After the optimizer update,
+    it synchronizes the parameters on the different nodes using SlowMo
+    (https://arxiv.org/abs/1910.00643).
+    The batch size should be larger than the number of GPUs used locally.
+    Creation of this class requires that ``torch.distributed`` to be already
+    initialized, by calling :func:`torch.distributed.init_process_group`.
+
+    To use ``DistributedDataParallel`` on a host with N GPUs, you should spawn
```
[the remark this answers is not in the document]
Comment (vtantia) [reply fairscale/nn/data_parallel/gossip/distributed.py:69]: Should we point them to the DDP documentation in that case?

File: fairscale/nn/data_parallel/gossip/utils/helpers.py (hunk 9 above)
Comment (vtantia) [reply fairscale/nn/data_parallel/gossip/utils/helpers.py:100]: This is being used in sgp_gossip_target. I've made it a future TODO to get rid of it

File: fairscale/nn/data_parallel/gossip/distributed.py (hunk 1 above)
Comment (vtantia) [reply fairscale/nn/data_parallel/gossip/distributed.py:142]: In those cases, how did you choose 8-12 (was it through tuning)? I had added this based on my experience with tuning it on Imagenet for vision, and the WMT 16 En-De dataset for translation, but can definitely change this

File: fairscale/nn/data_parallel/gossip/distributed.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mikerabbat) [reply fairscale/nn/data_parallel/gossip/distributed.py:69]: It just seems strange to me because this is the documentation for GossipDataParallel (or SlowmoDataParallel), so why are we talking about how to use `DistributedDataParallel` here?

File: fairscale/nn/data_parallel/gossip/distributed.py (hunk 1 above)
Comment (mikerabbat) [reply fairscale/nn/data_parallel/gossip/distributed.py:142]: Yes, based on experimenting/tuning. Basically, increasing frequency has diminishing returns in terms of how much time is saved, and increasing it can also degrade performance. So using smaller values sometimes will give better accuracy without really changing the runtime much.

File: fairscale/nn/data_parallel/gossip/distributed.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vtantia) [reply fairscale/nn/data_parallel/gossip/distributed.py:69]: Ahh I see the point you mean. I didn't get the meaning of copy-pasta haha

File: fairscale/nn/data_parallel/gossip/distributed.py [hunk 11]
```diff
@@ -146,7 +284,7 @@ def __init__(
 
         # slowmo being set to False is equivalent to slowmo_lr being set to 1 and slowmo_momentum being set to 0
         # This condition is ensuring the values are safe to use even when slowmo is disabled
-        self.slowmo = False if slowmo_lr == 1 and slowmo_momentum == 0 else slowmo
+        self.slowmo = False if slowmo_lr == 1 and slowmo_momentum == 0 else True
```
Comment (odelalleau) [inline fairscale/nn/data_parallel/gossip/distributed.py:287]:
```suggestion
        self.slowmo = slowmo_lr != 1 or slowmo_momentum != 0
```

File: fairscale/nn/data_parallel/gossip/distributed.py (hunk 4 above)
Comment (vtantia) [reply fairscale/nn/data_parallel/gossip/distributed.py:102]: I shall explain this more in the comment for slowmo_memory_efficient, a new parameter I have added to handle your second concern. That is in another branch and I send it to you for review.

The default was chosen by doing some preliminary tuning. With this value, the amortized time taken by SlowMo is quite negligible (on the order of a few milliseconds) so we went ahead with this value. We can revisit this number in the future if the need arises

Comment (mikerabbat) [reply fairscale/nn/data_parallel/gossip/distributed.py:102]: I think it's fine to leave it as a default and let a user adjust it as they see fit.

**Review by mikerabbat — APPROVED** [review]
Looks good!

**Comment by vtantia** [comment]
Merged into slowmo
